api/src/main/java/org/openmrs/api/impl/MedicationDispenseServiceImpl.java

```diff
@@ -0,0 +1,95 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ * 
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.api.impl;
+
+import org.apache.commons.lang.BooleanUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.openmrs.MedicationDispense;
+import org.openmrs.api.APIException;
+import org.openmrs.api.MedicationDispenseService;
+import org.openmrs.api.context.Context;
+import org.openmrs.api.db.MedicationDispenseDAO;
+import org.openmrs.parameter.MedicationDispenseCriteria;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.Date;
+import java.util.List;
+
+/**
+ * This class implements the {@link MedicationDispenseService} interface
+ * It defines the API for interacting with MedicationDispense objects.
+ * @since 2.6.0
+ */
+@Transactional
+public class MedicationDispenseServiceImpl extends BaseOpenmrsService implements MedicationDispenseService {
+
+	private MedicationDispenseDAO medicationDispenseDAO;
+
+	public void setMedicationDispenseDAO(MedicationDispenseDAO conditionDAO) {
+		this.medicationDispenseDAO = conditionDAO;
+	}
+
+	public MedicationDispenseDAO getMedicationDispenseDAO() {
+		return medicationDispenseDAO;
+	}
+
+	@Override
+	@Transactional(readOnly = true)
+	public MedicationDispense getMedicationDispense(Integer medicationDispenseId) {
+		return medicationDispenseDAO.getMedicationDispense(medicationDispenseId);
+	}
+
+	@Override
+	@Transactional(readOnly = true)
+	public MedicationDispense getMedicationDispenseByUuid(String uuid) {
+		return medicationDispenseDAO.getMedicationDispenseByUuid(uuid);
+	}
+
+	@Override
+	@Transactional(readOnly = true)
+	public List<MedicationDispense> getMedicationDispenseByCriteria(MedicationDispenseCriteria criteria) {
+		return medicationDispenseDAO.getMedicationDispenseByCriteria(criteria);
+	}
+
+	@Override
+	public MedicationDispense saveMedicationDispense(MedicationDispense medicationDispense) {
+		// TODO: Do we want to void and re-create if represents an edit with changed data?
```

Comment:
@mogoodrich / @dkayiwa / @ibacher - question here.  Do we want to have some logic in the save method of either the service or DAO that enforces / magically voids and recreates on edit, or not?  We don't do this for most data, we do it most comprehensively for Obs.  I see some support for this in the Condition Service, but I also see some things that don't look quite right in that domain that I wouldn't want to replicate (i.e. I don't want to void and recreate if the MedicationDispense hasn't changed).  Possibly there is some dirty checking we can do in the DAO layer or with Hibernate Intercepting, but not sure if there are patterns established we can follow.  We should discuss.

Reply:
My gut would be no, but interested in others thoughts and adding @bmamlin .  Beyond Obs, don't we do so for Orders.  @bmamlin  @ibacher  does FHIR provide any guidance around this (not that I'm saying we 100% should follow what FHIR does, but interested in their thoughts toward it).

Reply:
OK, for now I'm going to leave this as-is, not not explicitly void and recreate, as it adds significant complexity. Happy to continue to discuss.

Reply:
Well, the reason for the void-and-recreate behaviour is to be sure we keep a historical record of previous values. FHIR has a different approach to it, which is, essentially, that every resource is versioned. So, e.g. if I request `/MedicationRequest/123`, I get the latest version of it, but if I request `/MedicationRequest/123/_history/2` I might get a previous version of it (at least assuming the FHIR server stores the previous version of the resource).

Reply:
Yeah, my gut is that we _should_ do this, but it would be great if there were some DAO utilities that the service layer could use to determine if a given object is dirty or not.  This object has like 35 properties, so doing a comparison (and maintaining it as properties change) feels a bit cumbersome and hard to maintain.  Thoughts @ibacher ?

Reply:
Well there's what we do in Obs with a boolean `dirty` field and then every setter calls [`markAsDirty()`](https://github.com/openmrs/openmrs-core/blob/ed5eec88b4d3bfae584d9e6941ca6fb45cd54648/api/src/main/java/org/openmrs/Obs.java#L1198)...

It could be somewhat optimized by just making sure the `dirty` field defaults to `false`, e.g.,:

```java
@Transient
private transient boolean dirty = false;
```

Which would hopefully enable us to do away with the `ObsPostLoadEventListener`.

We could potentially use code generation to do the same thing (if you run Hibernate's newer "bytecode-enhanced dirty tracking", this is actually what it uses internally to check the state).

Reply:
Talking with @mogoodrich more, there are differences in intent behind MedicationDispense and Obs.  Obs typically, if you are editing them, means that you are correcting a mistake.  With MedicationDispense, there are several fields that seem like they'd get set progressively as you work - i.e. setting status, preparedBy, datePrepared, dateHandedOver, etc.  I don't know if there is value in retaining all of the permutations of these as voided data.